File: Backend/auth-service/src/main/java/com/tms/authservice/service/AuthServiceImpl.java
package com.tms.authservice.service;

import java.util.List;

import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import com.tms.authservice.dto.ChangePasswordRequest;
import com.tms.authservice.dto.CurrentUserResponse;
import com.tms.authservice.dto.LoginRequest;
import com.tms.authservice.dto.LoginResponse;
import com.tms.authservice.dto.RegisterHrRequest;
import com.tms.authservice.dto.UserLookupResponse;
import com.tms.authservice.entity.User;
import com.tms.authservice.entity.enums.AccountStatus;
import com.tms.authservice.entity.enums.ApprovalStatus;
import com.tms.authservice.entity.enums.Role;
import com.tms.authservice.repository.UserRepository;

import com.tms.authservice.security.JwtService;

@Service
public class AuthServiceImpl implements AuthService {

	private final UserRepository userRepository;
	private final BCryptPasswordEncoder passwordEncoder;
	private final JwtService jwtService;

    public AuthServiceImpl(UserRepository userRepository,
            BCryptPasswordEncoder passwordEncoder,
            JwtService jwtService) {

this.userRepository = userRepository;
this.passwordEncoder = passwordEncoder;
this.jwtService = jwtService;
}
    


    //=============================================
    //          GENERIC USER REGISTRATION
    //=============================================
    @Override
    public String registerUser(RegisterHrRequest request,
                               Role role,
                               ApprovalStatus approvalStatus,
                               AccountStatus accountStatus) {

        // Check duplicate username
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Username already exists");
        }

        // Check duplicate email
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }

        User user = User.builder()
                .username(request.getUsername())
                .password(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .email(request.getEmail())
                .contact(request.getContact())
                .joiningDate(request.getJoiningDate())
                .role(role)
                .approvalStatus(approvalStatus)
                .accountStatus(accountStatus)
                .build();

        userRepository.save(user);

        return role + " Registered Successfully";
    }

    //=============================================
    //                LOGIN
    //=============================================
    @Override
    public AuthenticatedLogin login(LoginRequest request) {

        User user = userRepository
                .findByUsernameAndApprovalStatusAndAccountStatus(
                        request.getUsername(),
                        ApprovalStatus.APPROVED,
                        AccountStatus.ACTIVE)
                .orElseThrow(() ->
                        new RuntimeException("Invalid Username or Account Not Active"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid Password");
        }

        // Generate JWT Token
        String token = jwtService.generateToken(user);

        LoginResponse response = LoginResponse.builder()
                .userId(user.getUserId())
                .username(user.getUsername())
                .role(user.getRole().name())
                .message("Login Successful")
                .build();

        // The controller places the token only in an HttpOnly cookie.
        return new AuthenticatedLogin(response, token);
    }

    //=============================================
    //       RESTORE AUTHENTICATED USER SESSION
    //=============================================
    @Override
    public CurrentUserResponse getCurrentUser(String username) {

        // Re-check the database so an inactive or unapproved account is not
        // restored only because an older JWT cookie is still present.
        User user = userRepository
                .findByUsernameAndApprovalStatusAndAccountStatus(
                        username,
                        ApprovalStatus.APPROVED,
                        AccountStatus.ACTIVE)
                .orElseThrow(() ->
                        new RuntimeException("User account is not active"));

        // Return only the information required by Redux and route authorization.
        return CurrentUserResponse.builder()
                .userId(user.getUserId())
                .username(user.getUsername())
                .role(user.getRole().name())
                .build();
    }

    //=============================================
    //       ACTIVE MANAGER/EMPLOYEE LOOKUPS
    //=============================================
    @Override
    public List<UserLookupResponse> getActiveManagers() {

        return getActiveUsersByRole(Role.MANAGER);
    }

    @Override
    public List<UserLookupResponse> getActiveEmployees() {

        return getActiveUsersByRole(Role.EMPLOYEE);
    }

    /**
     * Returns only approved and active users. This prevents HR from assigning
     * an inactive account to a Project.
     */
    private List<UserLookupResponse> getActiveUsersByRole(Role role) {

        return userRepository
                .findByRoleAndApprovalStatusAndAccountStatusOrderByFirstNameAscLastNameAsc(
                        role,
                        ApprovalStatus.APPROVED,
                        AccountStatus.ACTIVE)
                .stream()
                .map(this::toLookupResponse)
                .toList();
    }

    /**
     * Converts the User entity to a safe DTO. The password is never returned.
     */
    private UserLookupResponse toLookupResponse(User user) {

        String fullName =
                (user.getFirstName() + " " + user.getLastName()).trim();

        return UserLookupResponse.builder()
                .userId(user.getUserId())
                .username(user.getUsername())
                .fullName(fullName)
                .email(user.getEmail())
                .build();
    }

    //=============================================
    //          GET ALL PENDING HR
    //=============================================
    @Override
    public List<User> getPendingHrRequests() {

        return userRepository.findByRoleAndApprovalStatus(
                Role.HR_HEAD,
                ApprovalStatus.PENDING);
    }

    //=============================================
    //             APPROVE HR
    //=============================================
    @Override
    public String approveHr(Integer userId) {

        User user = userRepository.findById(userId)
                .orElseThrow(() ->
                        new RuntimeException("HR Not Found"));

        user.setApprovalStatus(ApprovalStatus.APPROVED);
        user.setAccountStatus(AccountStatus.ACTIVE);

        userRepository.save(user);

        return "HR Approved Successfully";
    }

    //=============================================
    //             REJECT HR
    //=============================================
    @Override
    public String rejectHr(Integer userId) {

        User user = userRepository.findById(userId)
                .orElseThrow(() ->
                        new RuntimeException("HR Not Found"));

        user.setApprovalStatus(ApprovalStatus.REJECTED);
        user.setAccountStatus(AccountStatus.INACTIVE);

        userRepository.save(user);

        return "HR Rejected Successfully";
    }
    
    
  //=============================================
  //             REST PASSWORD 
  //=============================================
    @Override
    public void changePassword(ChangePasswordRequest request) {

        Authentication authentication = SecurityContextHolder
                .getContext()
                .getAuthentication();

        String username = authentication.getName();

        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("User not found."));

        // Verify current password
        if (!passwordEncoder.matches(request.getOldPassword(), user.getPassword())) {
            throw new IllegalArgumentException("Current password is incorrect.");
        }

        // Verify new password and confirm password
        if (!request.getNewPassword().equals(request.getConfirmPassword())) {
            throw new IllegalArgumentException("New password and confirm password do not match.");
        }

        // Prevent using the same password
        if (passwordEncoder.matches(request.getNewPassword(), user.getPassword())) {
            throw new IllegalArgumentException("New password must be different from the current password.");
        }

        // Encode and save new password
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));

        userRepository.save(user);
    }

}

File: Backend/DotNet env/WorkPlus.AuthService/Services/AuthServiceImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using BCrypt.Net;
using Microsoft.EntityFrameworkCore;
using WorkPlus.AuthService.Data;
using WorkPlus.AuthService.DTOs;
using WorkPlus.AuthService.Models;
using WorkPlus.Shared.Enums;
using WorkPlus.Shared.Security;

namespace WorkPlus.AuthService.Services
{
    public class AuthServiceImpl : IAuthService
    {
        private readonly AuthDbContext _context;

        public AuthServiceImpl(AuthDbContext context)
        {
            _context = context;
        }

        public async Task<string> RegisterUserAsync(RegisterHrRequest request, Role role, ApprovalStatus approvalStatus, AccountStatus accountStatus)
        {
            if (await _context.Users.AnyAsync(u => u.Username == request.Username))
            {
                throw new InvalidOperationException("Username is already taken");
            }

            if (await _context.Users.AnyAsync(u => u.Email == request.Email))
            {
                throw new InvalidOperationException("Email is already registered");
            }

            var user = new User
            {
                Username = request.Username,
                Password = BCrypt.Net.BCrypt.HashPassword(request.Password),
                FirstName = request.FirstName,
                LastName = request.LastName,
                Email = request.Email,
                Contact = request.Contact,
                Role = role,
                ApprovalStatus = approvalStatus,
                AccountStatus = accountStatus,
                JoiningDate = request.JoiningDate,
                CreatedAt = DateTime.UtcNow
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            return role == Role.HR_HEAD && approvalStatus == ApprovalStatus.PENDING
                ? "HR registration submitted successfully. Awaiting Admin approval."
                : "User registered successfully.";
        }

        public async Task<(User? User, string? Token)> LoginAsync(LoginRequest request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Username == request.Username);
            if (user == null)
            {
                return (null, null);
            }

            bool isPasswordValid = false;
            if (user.Password.StartsWith("$2a$") || user.Password.StartsWith("$2b$") || user.Password.StartsWith("$2y$"))
            {
                try
                {
                    isPasswordValid = BCrypt.Net.BCrypt.Verify(request.Password, user.Password);
                }
                catch
                {
                    isPasswordValid = false;
                }
            }
            else
            {
                // Fallback for sample DB seed plain-text passwords
                isPasswordValid = request.Password == user.Password;
            }

            if (!isPasswordValid)
            {
                return (null, null);
            }

            if (user.ApprovalStatus != ApprovalStatus.APPROVED || user.AccountStatus != AccountStatus.ACTIVE)
            {
                return (null, null);
            }

            var token = JwtTokenHelper.GenerateToken(user.Username, user.Role.ToString());
            return (user, token);
        }

        public async Task<CurrentUserResponse?> GetCurrentUserAsync(string username)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Username == username);
            if (user == null) return null;

            return new CurrentUserResponse
            {
                UserId = user.UserId,
                Username = user.Username,
                Email = user.Email,
                Role = user.Role.ToString(),
                FirstName = user.FirstName,
                LastName = user.LastName,
                ApprovalStatus = user.ApprovalStatus.ToString(),
                AccountStatus = user.AccountStatus.ToString()
            };
        }

        public async Task<List<UserLookupResponse>> GetActiveManagersAsync()
        {
            return await _context.Users
                .Where(u => u.Role == Role.MANAGER && u.ApprovalStatus == ApprovalStatus.APPROVED && u.AccountStatus == AccountStatus.ACTIVE)
                .Select(u => new UserLookupResponse
                {
                    UserId = u.UserId,
                    Username = u.Username,
                    FullName = $"{u.FirstName} {u.LastName}",
                    Email = u.Email,
                    Role = u.Role.ToString()
                }).ToListAsync();
        }

        public async Task<List<UserLookupResponse>> GetActiveEmployeesAsync()
        {
            return await _context.Users
                .Where(u => u.Role == Role.EMPLOYEE && u.ApprovalStatus == ApprovalStatus.APPROVED && u.AccountStatus == AccountStatus.ACTIVE)
                .Select(u => new UserLookupResponse
                {
                    UserId = u.UserId,
                    Username = u.Username,
                    FullName = $"{u.FirstName} {u.LastName}",
                    Email = u.Email,
                    Role = u.Role.ToString()
                }).ToListAsync();
        }

        public async Task<List<User>> GetPendingHrRequestsAsync()
        {
            return await _context.Users
                .Where(u => u.Role == Role.HR_HEAD && u.ApprovalStatus == ApprovalStatus.PENDING)
                .ToListAsync();
        }

        public async Task<string> ApproveHrAsync(int userId)
        {
            var user = await _context.Users.FindAsync(userId);
            if (user == null) throw new KeyNotFoundException("User not found");

            user.ApprovalStatus = ApprovalStatus.APPROVED;
            user.AccountStatus = AccountStatus.ACTIVE;
            await _context.SaveChangesAsync();
            return "HR approved successfully.";
        }

        public async Task<string> RejectHrAsync(int userId)
        {
            var user = await _context.Users.FindAsync(userId);
            if (user == null) throw new KeyNotFoundException("User not found");

            user.ApprovalStatus = ApprovalStatus.REJECTED;
            user.AccountStatus = AccountStatus.INACTIVE;
            await _context.SaveChangesAsync();
            return "HR rejected successfully.";
        }

        public async Task ChangePasswordAsync(ChangePasswordRequest request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Username == request.Username);
            if (user == null) throw new UnauthorizedAccessException("Invalid credentials");

            bool isOldValid = false;
            if (user.Password.StartsWith("$2a$") || user.Password.StartsWith("$2b$") || user.Password.StartsWith("$2y$"))
            {
                isOldValid = BCrypt.Net.BCrypt.Verify(request.OldPassword, user.Password);
            }
            else
            {
                isOldValid = request.OldPassword == user.Password;
            }

            if (!isOldValid) throw new UnauthorizedAccessException("Invalid credentials");

            user.Password = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);
            await _context.SaveChangesAsync();
        }
    }
}
